src/main/java/org/spongepowered/common/launch/transformer/tracker/TrackerClassTransformer.java

```diff
@@ -0,0 +1,210 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.launch.transformer.tracker;
+
+import static org.spongepowered.asm.lib.Opcodes.ACC_PRIVATE;
+import static org.spongepowered.asm.lib.Opcodes.ACC_STATIC;
+import static org.spongepowered.asm.lib.Opcodes.ALOAD;
+import static org.spongepowered.asm.lib.Opcodes.ARETURN;
+import static org.spongepowered.asm.lib.Opcodes.ASM5;
+import static org.spongepowered.asm.lib.Opcodes.CHECKCAST;
+import static org.spongepowered.asm.lib.Opcodes.DLOAD;
+import static org.spongepowered.asm.lib.Opcodes.DRETURN;
+import static org.spongepowered.asm.lib.Opcodes.FLOAD;
+import static org.spongepowered.asm.lib.Opcodes.FRETURN;
+import static org.spongepowered.asm.lib.Opcodes.F_SAME;
+import static org.spongepowered.asm.lib.Opcodes.IFEQ;
+import static org.spongepowered.asm.lib.Opcodes.ILOAD;
+import static org.spongepowered.asm.lib.Opcodes.INSTANCEOF;
+import static org.spongepowered.asm.lib.Opcodes.INVOKEINTERFACE;
+import static org.spongepowered.asm.lib.Opcodes.INVOKESTATIC;
+import static org.spongepowered.asm.lib.Opcodes.INVOKEVIRTUAL;
+import static org.spongepowered.asm.lib.Opcodes.IRETURN;
+import static org.spongepowered.asm.lib.Opcodes.LLOAD;
+import static org.spongepowered.asm.lib.Opcodes.LRETURN;
+import static org.spongepowered.asm.lib.Opcodes.RETURN;
+
+import net.minecraft.launchwrapper.IClassTransformer;
+import org.spongepowered.asm.lib.ClassReader;
+import org.spongepowered.asm.lib.ClassWriter;
+import org.spongepowered.asm.lib.Label;
+import org.spongepowered.asm.lib.MethodVisitor;
+import org.spongepowered.asm.lib.Type;
+import org.spongepowered.asm.lib.tree.AbstractInsnNode;
+import org.spongepowered.asm.lib.tree.ClassNode;
+import org.spongepowered.asm.lib.tree.MethodInsnNode;
+import org.spongepowered.asm.lib.tree.MethodNode;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.Set;
+
+public class TrackerClassTransformer implements IClassTransformer {
+
+    @Override
+    public byte[] transform(String name, String transformedName, byte[] basicClass) {
+        if (TrackerRegistry.TRACKER_CLASSES.contains(name)) {
+            return basicClass;
+        }
+
+        final ClassReader classReader = new ClassReader(basicClass);
+
+        final ClassNode classNode = new ClassNode(ASM5);
+        classReader.accept(classNode, ClassReader.EXPAND_FRAMES);
+
+        final Set<String> addedMethods = new HashSet<>();
+        boolean change = false;
+        for (MethodNode methodNode : new ArrayList<>(classNode.methods)) {
+            if (processMethodNode(classNode, methodNode, addedMethods)) {
```

Comment:
I started with a tree, but ended up not needing it. I will change it to a ClassVisitor.